### PR TITLE
Rework concurrency limits using the number provided in Cloud API

### PR DESCRIPTION
**Pull request by Vlatombe**
Downstream PR for https://github.com/jenkinsci/jenkins/pull/4922

* Improves the `KubernetesCloud` implementation by using the provided count of planned nodes for the current label. This makes the enforcement of pod template limits exact. Since we don't have access to the number of planned nodes for all node provisioners, the global limit can be affected by an error margin.
Reworked the limits check algorithm to execute exactly one kubernetes api call per `KubernetesCloud#provision` call. The previous version was executing 2 calls for each potential node to provision.
* Introduces a `launching` field on `KubernetesComputer`. This intends to capture the moment where agents exist in kubernetes but not yet usable so that we can compute more accurately the limits.
* Reverts some of the `NoDelayProvisionerStategy` changes from #824. These changes were trying to avoid the standard strategy to run because this was overprovisioning the cloud by the number of planned nodes.

### CODE REVIEW

**Comment by chwehrli** [comment]
Looking forward for this PR to be merged as I believe there's a bug in the current implementation of KubernetesCloud.java line 614:

`        // check template-level concurrency limit using template-level labels
        Map<String, String> templateLabels = new HashMap<>(podLabels);
        templateLabels.putAll(template.getLabelsMap());
        numRunningOrPending = getNumActiveSlavePods(client, templateNamespace, podLabels);`

..should be('templateLabels' instead of 'podLabels'):
`        // check template-level concurrency limit using template-level labels
        Map<String, String> templateLabels = new HashMap<>(podLabels);
        templateLabels.putAll(template.getLabelsMap());
        numRunningOrPending = getNumActiveSlavePods(client, templateNamespace, templateLabels);`

..right?

In my Jenkins environment I somehow just don't manage to come up with a proper definition for POD limits - looks like no matter what limit I set on the various template definitions the overall number of PODs is kind of limited to the lowest number.

Kind regards,
Christian

**Comment by schottsfired** [comment]
That's correct @chwehrli, I contributed a new bug ([JENKINS-63075](https://issues.jenkins-ci.org/browse/JENKINS-63705) via https://github.com/jenkinsci/kubernetes-plugin/pull/824 which shipped in v1.27.1. 

Sorry about that!

https://github.com/jenkinsci/kubernetes-plugin/pull/843 is an attempt to craft an emergency fix (read: revert), but this PR is the actual fix for the original issue, [JENKINS-59959](https://issues.jenkins-ci.org/browse/JENKINS-59959).

**Review by schottsfired — APPROVED** [review]
Tests 👍 

Nice PRs @Vlatombe !

File: pom.xml [hunk 1]
```diff
@@ -252,7 +252,7 @@
     <dependencies>
       <dependency>
         <groupId>io.jenkins.tools.bom</groupId>
-        <artifactId>bom-2.222.x</artifactId>
+        <artifactId>bom-2.222.x</artifactId> <!-- TODO https://github.com/jenkinsci/bom/pull/339 -->
         <version>12</version>
```
Comment (jglick) [inline pom.xml:256]: New BOM is ready; also see #875.

**Comment by Vlatombe** [comment]
time for a final look before I merge this.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java [hunk 2]
```diff
@@ -527,31 +527,49 @@ public KubernetesClient connect() throws KubernetesAuthException, IOException {
...
+                            " launching: {4}",
+                    new Object[] {label, toBeProvisioned, currentExecutorsCount, plannedCapacity, launchingExecutorsCount});
+            List<NodeProvisioner.PlannedNode> plannedNodes = new ArrayList<>();
+            List<Pod> pods = getPodsWithLabels(namespace, getPodLabelsMap());
+            if (pods != null) {
+                // Count planned and executors not yet launching (e.g. everything that is not in kubernetes yet)
+                int plannedCount = plannedCapacity + currentExecutorsCount - launchingExecutorsCount;
+                int remainingGlobalSlots = getRemainingGlobalSlots(pods, plannedCount);
+                if (remainingGlobalSlots > 0) {
+                    LOGGER.log(Level.FINE, "Global slots left for \"{0}\": {1}", new Object[]{name, remainingGlobalSlots});
```
Comment (jglick) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesCloud.java:552]:
```suggestion
                    LOGGER.fine(() -> "Global slots left for " + name + ": " + remainingGlobalSlots);
```

if you like

**Review by jglick — APPROVED** [review]
Do not know anything about the actual logic here. Metadata looks fine.

**Comment by clabu** [comment]
It looks like this PR will fix https://issues.jenkins-ci.org/browse/JENKINS-64048, right? Thanks!

**Comment by schottsfired** [comment]
Hi @clabu, sorry if my comment was confusing. This PR fixes [JENKINS-59959](https://issues.jenkins-ci.org/browse/JENKINS-59959). Version 1.27.1 contained a bug ([JENKINS-63705](https://issues.jenkins-ci.org/browse/JENKINS-63705)) caused by https://github.com/jenkinsci/kubernetes-plugin/pull/824 (my failed attempt to fix JENKINS-59959), and the bug was removed in 1.27.1.1 via https://github.com/jenkinsci/kubernetes-plugin/pull/844.

Not sure if this PR fixes [JENKINS-64048](https://issues.jenkins-ci.org/browse/JENKINS-64048) because the issue tracker is down at the moment.

**Comment by kay-ramme** [comment]
still see this .. 
Jenkins: 2.267 
K8S plugin: 1.28.2

**Comment by Vlatombe** [comment]
@kay-ramme can you be more explicit? There are some limitations we'll never be able to overcome.
> Since we don't have access to the number of planned nodes for all node provisioners, the global limit can be affected by an error margin.

**Comment by BenjaminITS** [comment]
Hello,

I still have this problem with Jenkins 2.267, plugin 1.28.3 and bare metal k8s 1.15.12. 
The concurrency limit works at the beginning but as soon as a slot is released all the queued jobs are launched. 

> Since we don't have access to the number of planned nodes for all node provisioners, the global limit can be affected by an error margin.

Is that what you're talking about?

**Pull request merged**